[AIRFLOW-6128] Simplify plugins_manager and webserver plugin code

potiuk opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-6128

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

While adding type annotations, it was apparent that the code for
plugin manager and webserver initialisation was rather convoluted
and complex. This change fixes it as follow up after untangling
cyclic dependnencies.

As a side effect - it turned out that the test plugin view
in airflow RBAC implementation was not working since 1.10.1 - it threw "render
is not an attribute" error - this has been fixed in this commit
as well. The Plugins/TestView now works as expected. See
https://stackoverflow.com/questions/53701030/airflow-plugins-rbac-enabled-blueprint-not-working?answertab=active#tab-top

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

ashb commented on airflow/config_templates/airflow_local_settings.py:
```diff
@@ -201,17 +199,17 @@
                 'class': 'airflow.utils.log.es_task_handler.ElasticsearchTaskHandler',
                 'formatter': 'airflow',
-                'base_log_folder': os.path.expanduser(BASE_LOG_FOLDER),
+                'base_log_folder': str(os.path.expanduser(BASE_LOG_FOLDER)),
```
The changes in this file seem to be just type hint related, so should be unrelated to this change?

potiuk replied:
I split it out to separate commit.

potiuk committed:
[AIRFLOW-6128] Simplify plugins_manager and webserver plugin code

While adding type annotations, it was apparent that the code for
plugin manager and webserver initialisation was rather convoluted
and complex. This change fixes it as follow up after untangling
cyclic dependnencies.

As a side effect - it turned out that the test plugin view
in airflow RBAC implementation was not working since 1.10.1 - it threw "render
is not an attribute" error - this has been fixed in this commit
as well. The Plugins/TestView now works as expected. See
https://stackoverflow.com/questions/53701030/airflow-plugins-rbac-enabled-blueprint-not-working?answertab=active#tab-top

feluelle commented on airflow/plugins_manager.py:
```diff
@@ -159,161 +143,195 @@ def is_valid_plugin(plugin_obj, existing_plugins):
-            if file_ext != '.py':
-                continue
+def import_plugin(filepath: str, mod_name: str, root: str) -> None:
+    """Imports plugin module."""
```
Don't you think a full documentation is needed here?

feluelle commented on airflow/plugins_manager.py:
```diff
@@ -159,161 +143,195 @@ def is_valid_plugin(plugin_obj, existing_plugins):
-            log.error('Failed to import plugin %s', path)
-            import_errors[path] = str(e)
+def try_to_import_plugin(filepath: str, root: str) -> None:
+    """Tries to import plugin from the file specified. Skips it if not a valid python file."""
```
Same here.

feluelle commented on airflow/plugins_manager.py:
```diff
@@ -159,161 +143,195 @@ def is_valid_plugin(plugin_obj, existing_plugins):
+
+def find_airflow_plugins() -> None:
+    """Crawl through the plugins folder to find AirflowPlugin derivatives"""
+    assert settings.PLUGINS_FOLDER, "Plugins folder is not set"
```
I thought we did decide to not use `assert` for raising exceptions.

feluelle commented on airflow/plugins_manager.py:
```diff
@@ -159,161 +143,195 @@ def is_valid_plugin(plugin_obj, existing_plugins):
+def get_stat_name_handler() -> Optional[Callable[[str], str]]:
+    """Retrieves the single configured stats_handler (there can be only one!)"""
+    assert len(stat_name_handlers) in {0, 1}, \
+        f'Specified more than one stat_name_handler ({stat_name_handlers}) which is not allowed.'
```
> I thought we did decide to not use assert for raising exceptions.

Same here.

feluelle commented on airflow/plugins_manager.py:
```diff
@@ -159,161 +143,195 @@ def is_valid_plugin(plugin_obj, existing_plugins):
-        'Specified more than one stat_name_handler ({}) '
-        'is not allowed.'.format(stat_name_handlers))
+def get_stat_name_handler() -> Optional[Callable[[str], str]]:
+    """Retrieves the single configured stats_handler (there can be only one!)"""
```
Can you add the documentation for the return value, please?

feluelle commented on airflow/www/app.py:
```diff
@@ -89,167 +176,116 @@ def create_app(config=None, session=None, testing=False, app_name="Airflow"):
-
-    return app, appbuilder
+def configure_app(app_name: str, config: Optional[Dict[str, Union[str, bool]]], testing: bool):
+    """Configure the application."""
```
Please add a description to the parameters.

feluelle commented on airflow/www/app.py:
```diff
@@ -89,167 +176,116 @@ def create_app(config=None, session=None, testing=False, app_name="Airflow"):
 
-def root_app(env, resp):
+def root_app(_, resp: Callable[[bytes, List[Tuple[str, str]]], None]):
+    """Root application function - returns 404 if the path is not found."""
```
Same here.

feluelle commented on airflow/www/app.py:
```diff
@@ -89,167 +176,116 @@ def create_app(config=None, session=None, testing=False, app_name="Airflow"):
+def cached_app(config: Optional[Dict[str, Union[str, bool]]] = None,
+               session: Optional[Session] = None,
+               testing: bool = False) -> Flask:
+    """Returns cached application if it's there. Creates it otherwise."""
```
And here.

feluelle commented on airflow/www/app.py:
```diff
@@ -89,167 +176,116 @@ def create_app(config=None, session=None, testing=False, app_name="Airflow"):
-    global appbuilder
+def cached_appbuilder(config: Optional[Dict[str, Union[str, bool]]] = None, testing=False) -> \
+        AppBuilder:
+    """Returns cached appbuilder."""
```
And here.
